File: ruby/text/match_json.rb
# encoding: UTF-8
#
# Example script to issue a match.json request
#
require 'rubygems'
require 'uri'
require 'net/http'
require 'digest/md5'
require 'openssl'
require 'base64'
require 'time'
require 'zlib'
require 'json'

# Set your environment variables to the keys obtained from https://www.idilia.com/developer/my-projects
accessKey = ENV['IDILIA_ACCESS_KEY'] || ""
privateKey = ENV['IDILIA_PRIVATE_KEY'] || ""
raise "You have to set environment variables IDILIA_ACCESS_KEY and IDILIA_PRIVATE_KEY" if accessKey.empty? or privateKey.empty?

# The Tweets that we'll process
texts=[
  "RT @blecklerr: just saw a southern tide decal on a nissan with dark tint and the biggest shiniest rims. #theyreconfused #WhatsGoingOnHere",
  "In honor of the Crimson Tide, here's the song of the day. Welcome to Miami, vien bonito amiami !! See you at... http://t.co/vdkan4mN"
]
textMime="text/tweet; charset=UTF-8"

# Open up a connection that we can re-use for several requests
uri = URI('http://api.idilia.com/1/text/match.json')
Net::HTTP.start(uri.host, uri.port) do | http |

  texts.each do | text |
    
    # Compute the authorization
    date = Time.now.httpdate()
    md5 = Base64.encode64(Digest::MD5.digest(text)).chomp
    toSign=date + '-' + uri.host + '-' + uri.path + '-' + md5
    signature=Base64.encode64(OpenSSL::HMAC.digest('sha256', privateKey, toSign)).chomp
    authorization = "IDILIA " + accessKey + ":" + signature
  
    # Create the HTTP request with the headers and parameters
    request = Net::HTTP::Post.new(uri.path, {
      'Accept-Encoding' => 'gzip',
      'Authorization' => authorization,
      'Date' => date,
      'Host' => uri.host
      })
    request.set_form_data({
        'text' => text,
        'textMime' => textMime,
        'filter' => '{"fsk":"tide/N1"}',
        'requestId' => 'my-request'
      })
    request.set_content_type('application/x-www-form-urlencoded; charset=UTF-8')
  
    # Get the response and decode it
    http.read_timeout = 3600 + 60; # slightly exceeds default value for parameter "timeout"
    httpResponse = http.request(request)
    body = httpResponse.body
    body = Zlib::Inflate.new(16+Zlib::MAX_WBITS).inflate(body) if httpResponse['Content-Encoding'] == 'gzip'
  
    # Parse the response
    case httpResponse
    when Net::HTTPOK
      resp = JSON.parse(body)
      puts text
      if resp.has_key?("matches")
        resp["matches"].each do | m |
          if m.has_key?("foundSk")
            puts "  Found identify or equivalent sensekey #{m["foundSk"]} at offset #{m["position"][0]} with confidence #{m["conf"]} and for reason #{m["reasons"][0]}"
          else
            puts '  Word match but wrong sense'
          end
        end
      else
        puts "  No word match"
      end
    else
      puts body
      raise "  Some unexpected error occured when processing query [#{text}]"
    end
    puts
  end
end

File: python/text/match_json.py
#
# Example program for issuing match.json request
# processed immediately on response.
#

import urllib
import os
import httplib
import urlparse
import locale
import hashlib
import hmac, base64
import zlib
import json
from email.utils import formatdate
from datetime import datetime
from time import mktime

URL = "http://api.idilia.com/1/text/match.json"
ACCESS_KEY = os.environ.get("IDILIA_ACCESS_KEY")
PRIVATE_KEY = os.environ.get("IDILIA_PRIVATE_KEY")
if (ACCESS_KEY == None or PRIVATE_KEY == None or len(ACCESS_KEY) == 0 or len(PRIVATE_KEY) == 0):
  raise Exception("You have to set environment variables IDILIA_ACCESS_KEY and IDILIA_PRIVATE_KEY")


# Setup the HTTP connection. It can be re-used for multiple 
# consecutive requests as we use HTTP 1.1
url = urlparse.urlparse(URL)
httpConn = httplib.HTTPConnection(url.hostname, url.port, timeout=4000)
locale.setlocale(locale.LC_TIME, 'en_US')

# The Tweets that we'll process
texts=[
  "RT @blecklerr: just saw a southern tide decal on a nissan with dark tint and the biggest shiniest rims. #theyreconfused #WhatsGoingOnHere",
  "In honor of the Crimson Tide, here's the song of the day. Welcome to Miami, vien bonito amiami !! See you at... http://t.co/vdkan4mN"
]
textMime="text/tweet; charset=UTF-8"

for text in texts:
  # Compute the authorization
  date = formatdate(timeval=mktime(datetime.utcnow().timetuple()), localtime=True, usegmt=True)
  md5 = base64.b64encode(hashlib.md5(text).digest())
  toSign = "%s-%s-%s-%s" % (date, url.hostname, url.path, md5)
  signature = str(base64.b64encode(hmac.new(PRIVATE_KEY, toSign, hashlib.sha256).digest()))
  
  # The request parameters
  params = {
    'requestId' : 'mytest',
    'text' : text,
    'textMime' : textMime,
    'filter' : '{"fsk":"tide/N1"}'
  }
  
  # The HTTP headers that we use: Mostly for validation
  headers = {
      'Accept-Encoding': 'gzip',
      'Authorization' : "IDILIA %s:%s" % (ACCESS_KEY, signature),
      'Content-Type': "application/x-www-form-urlencoded",
      'Date' : date,
      'Host' : url.hostname
  }
  
  # Send the POST request and send it
  httpConn.request("POST", url.path, urllib.urlencode(params), headers)
  
  # Get the response and decode it
  httpResponse = httpConn.getresponse();
  httpBody = httpResponse.read();
  if httpResponse.getheader("Content-Encoding") == "gzip":
    httpBody = zlib.decompress(httpBody, 16+zlib.MAX_WBITS)
  
  # First validation in case something went wrong in the transport layer
  if httpResponse.status != httplib.OK:
      raise Exception(httpBody, httpResponse.status, httpResponse.reason)
  
  # Application layer validation
  resp = json.loads(httpBody)
  if resp.has_key('errorMsg'):
      raise Exception(resp['errorMsg'], resp['status'])
  
  # Display the senses found.
  print text;
  if 'matches' in resp:
    for m in resp['matches']:
      if 'foundSk' in m:
        print "  Found identify or equivalent sensekey %s at offset %d with confidence %3.2f and for reason %s" % (m["foundSk"], m["position"][0], m["conf"], m["reasons"][0])
      else:
        print "  Word match but wrong sense"
  else:
        print "  No word match"

  print
